packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,138 @@
-              severity: DiagnosticSeverity.Error,
-            },
-          ];
-        }
+function validateBody(schema: JSONSchema, target: any) {
+  const diagnostics = validateAgainstSchema(target, schema);
+  return diagnostics.length ? Either.left(diagnostics) : Either.right(target);
+}
 
-        if (encoding.style) {
-          const deserializer = body.get(encoding.style);
-          if (deserializer && schema.properties) {
-            const propertySchema = schema.properties[property];
-            target[property] = deserializer.deserialize(property, target, propertySchema as JSONSchema);
-          }
-        }
-      }
-    }
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
+  if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
+    const encodings = get(content, 'encodings', []);
+    const encodedUriParams = splitUriParams(target);
 
-    return validateAgainstSchema(target, schema).map(error =>
-      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
+    return pipe(
+      validateAgainstReservedCharacters(encodedUriParams, encodings),
+      Either.map(decodeUriEntities),
+      Either.map(deserializeFormBody.bind(undefined, schema, encodings)),
     );
   }
 
-  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
-    if (!mediaType) {
-      return specs[0];
-    }
+  return Either.right(target);
+}
 
-    const content = specs.find(c => c.mediaType === mediaType);
+function applyPrefix(prefix: string, diagnostics: IPrismDiagnostic[]): IPrismDiagnostic[] {
+  return diagnostics.map(d => ({ ...d, path: [prefix, ...(d.path || [])] }));
+}
 
-    if (!content) {
-      return specs[0];
-    }
+function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]) {
+  return pipe(
+    encodings,
+    Array.reduce<IHttpEncoding, IPrismDiagnostic[]>([], (diagnostics, encoding) => {
+      const allowReserved = get(encoding, 'allowReserved', false);
+      const property = encoding.property;
+      const value = encodedUriParams[property];
+
+      if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
+        diagnostics.push({
+          path: [property],
+          message: 'Reserved characters used in request body',
+          severity: DiagnosticSeverity.Error,
+        });
+      }
 
-    return content;
+      return diagnostics;
+    }),
+    diagnostics => (diagnostics.length ? Either.left(diagnostics) : Either.right(encodedUriParams)),
+  );
+}
+
+function deserializeFormBody(
+  schema: JSONSchema,
+  encodings: IHttpEncoding[],
+  decodedUriParams: Dictionary<string, string>,
+) {
+  if (!schema.properties) {
+    return decodedUriParams;
   }
+
+  return pipe(
+    Object.keys(schema.properties),
+    Array.reduce({}, (deserialized, property) => {
```

Comment:
can this reduce be mutation-free?

Reply:
That discussion may take lot of time, but let me explain my motivation.

Basically I consider this as a conflict between:
mutation-free vs. speed
or
perfectionism vs. rationality

Like in this example, the poor js engine will need to create a brand new object, assign all properties (we have no control over how many of those are there!), put another property and repeat those ops property-count times. 

I choose speed/rationality in small reducers over mutation-free/perfectionism. I think the reducer is small, well defined and mutating the value won't make any trouble.

Reply:
okey-dokey, but let's add least do this https://github.com/stoplightio/prism/pull/649#discussion_r329107225 ;)

Reply:
In general I am ok with such tradeoffs; messes will have to happen somewhere anyway but as long you can keep the stuff contained in one place where is not harming anybody but still a functional interface, I am ok with that.